File: ios/ContractRunner/Sources/ContractRunner/MediaIds.swift
import Foundation

public struct NuvioMediaId: Equatable {
    public let contentId: String
    public let videoId: String?
    public let isEpisode: Bool
    public let season: Int?
    public let episode: Int?
    public let kind: String
    public let addonLookupId: String

    public init(
        contentId: String,
        videoId: String?,
        isEpisode: Bool,
        season: Int?,
        episode: Int?,
        kind: String,
        addonLookupId: String
    ) {
        self.contentId = contentId
        self.videoId = videoId
        self.isEpisode = isEpisode
        self.season = season
        self.episode = episode
        self.kind = kind
        self.addonLookupId = addonLookupId
    }
}

public func normalizeNuvioMediaId(_ raw: String) -> NuvioMediaId {
    let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    if trimmed.isEmpty {
        return NuvioMediaId(
            contentId: "",
            videoId: nil,
            isEpisode: false,
            season: nil,
            episode: nil,
            kind: "content",
            addonLookupId: ""
        )
    }

    let suffix = parseEpisodeSuffix(trimmed)
    if let season = suffix.season,
       let episode = suffix.episode {
        let baseId = canonicalizeBaseId(suffix.baseId)
        if !baseId.isEmpty {
            let videoId = "\(baseId):\(season):\(episode)"
            return NuvioMediaId(
                contentId: baseId,
                videoId: videoId,
                isEpisode: true,
                season: season,
                episode: episode,
                kind: "episode",
                addonLookupId: videoId
            )
        }
    }

    let normalized = canonicalizeBaseId(stripSeriesPrefix(trimmed))

    return NuvioMediaId(
        contentId: normalized,
        videoId: nil,
        isEpisode: false,
        season: nil,
        episode: nil,
        kind: "content",
        addonLookupId: normalized
    )
}

public func formatIdForIdPrefixes(
    _ input: String,
    mediaType: String,
    idPrefixes: [String] = []
) -> String? {
    let raw = input.trimmingCharacters(in: .whitespacesAndNewlines)
    if raw.isEmpty {
        return nil
    }

    let suffix = parseEpisodeSuffix(raw)
    let baseRaw = suffix.baseId.trimmingCharacters(in: .whitespacesAndNewlines)
    if baseRaw.isEmpty {
        return nil
    }

    let normalizedBase = canonicalizeBaseId(baseRaw)
    let episodeSuffix: String
    if let season = suffix.season, let episode = suffix.episode {
        episodeSuffix = ":\(season):\(episode)"
    } else {
        episodeSuffix = ""
    }

    let providerKind = mediaType.caseInsensitiveCompare("movie") == .orderedSame ? "movie" : "show"

    var candidates: [String] = []
    var seen = Set<String>()

    if isImdbId(normalizedBase) {
        appendUnique("\(normalizedBase)\(episodeSuffix)", to: &candidates, seen: &seen)
        appendUnique("imdb:\(normalizedBase)\(episodeSuffix)", to: &candidates, seen: &seen)
        appendUnique("imdb:\(providerKind):\(normalizedBase)\(episodeSuffix)", to: &candidates, seen: &seen)
    }

    if let tmdbId = tmdbNumericIdOrNil(normalizedBase) {
        appendUnique("tmdb:\(tmdbId)\(episodeSuffix)", to: &candidates, seen: &seen)
        appendUnique("tmdb:\(providerKind):\(tmdbId)\(episodeSuffix)", to: &candidates, seen: &seen)
    }

    let needsNumericInference =
        isNumeric(normalizedBase) &&
        !isImdbId(normalizedBase) &&
        tmdbNumericIdOrNil(normalizedBase) == nil &&
        !containsSchemePrefix(normalizedBase)
    if needsNumericInference {
        let inferredProvider = inferNumericProvider(idPrefixes)
        if inferredProvider == nil && !idPrefixes.isEmpty {
            return nil
        }
        if let inferredProvider {
            appendUnique("\(inferredProvider):\(normalizedBase)\(episodeSuffix)", to: &candidates, seen: &seen)
            appendUnique("\(inferredProvider):\(providerKind):\(normalizedBase)\(episodeSuffix)", to: &candidates, seen: &seen)
        }
    }

    appendUnique("\(normalizedBase)\(episodeSuffix)", to: &candidates, seen: &seen)

    if !idPrefixes.isEmpty {
        for candidate in candidates {
            if idPrefixes.contains(where: { candidate.hasPrefix($0) }) {
                return candidate
            }
        }
        return nil
    }

    return candidates.first
}

private struct EpisodeSuffix {
    let baseId: String
    let season: Int?
    let episode: Int?
}

private func parseEpisodeSuffix(_ value: String) -> EpisodeSuffix {
    let normalized = stripSeriesPrefix(value.trimmingCharacters(in: .whitespacesAndNewlines))
    let parts = normalized.split(separator: ":").map(String.init)
    if parts.count < 3 {
        return EpisodeSuffix(baseId: normalized, season: nil, episode: nil)
    }

    guard let season = Int(parts[parts.count - 2]), season > 0,
          let episode = Int(parts[parts.count - 1]), episode > 0
    else {
        return EpisodeSuffix(baseId: normalized, season: nil, episode: nil)
    }

    let baseId = stripSeriesPrefix(parts.dropLast(2).joined(separator: ":").trimmingCharacters(in: .whitespacesAndNewlines))
    return EpisodeSuffix(baseId: baseId, season: season, episode: episode)
}

private func canonicalizeBaseId(_ value: String) -> String {
    let trimmed = stripSeriesPrefix(value.trimmingCharacters(in: .whitespacesAndNewlines))
    if trimmed.isEmpty {
        return ""
    }

    if let imdb = extractImdbId(trimmed) {
        return imdb
    }

    if let tmdb = extractTmdbId(trimmed) {
        return "tmdb:\(tmdb)"
    }

    return trimmed
}

private func stripSeriesPrefix(_ value: String) -> String {
    guard value.lowercased().hasPrefix("series:") else {
        return value
    }
    return String(value.dropFirst("series:".count))
}

private func extractImdbId(_ value: String) -> String? {
    let normalized = stripSeriesPrefix(value.trimmingCharacters(in: .whitespacesAndNewlines))
    if isImdbId(normalized) {
        return normalized.lowercased()
    }

    let parts = normalized.split(separator: ":").map(String.init)
    guard let first = parts.first, first.caseInsensitiveCompare("imdb") == .orderedSame else {
        return nil
    }

    guard let token = parts.reversed().first(where: { isImdbId($0) }) else {
        return nil
    }

    return token.lowercased()
}

private func extractTmdbId(_ value: String) -> String? {
    let normalized = stripSeriesPrefix(value.trimmingCharacters(in: .whitespacesAndNewlines))
    guard normalized.lowercased().hasPrefix("tmdb:") else {
        return nil
    }

    let parts = normalized.split(separator: ":").map(String.init)
    guard parts.count >= 2 else {
        return nil
    }

    if isNumeric(parts[1]) {
        return parts[1]
    }

    if parts.count > 2, isNumeric(parts[2]) {
        return parts[2]
    }

    return nil
}

private func tmdbNumericIdOrNil(_ value: String) -> String? {
    guard value.lowercased().hasPrefix("tmdb:") else {
        return nil
    }
    let remainder = String(value.dropFirst("tmdb:".count))
    let numeric = remainder.split(separator: ":", maxSplits: 1, omittingEmptySubsequences: false).first.map(String.init) ?? ""
    return isNumeric(numeric) ? numeric : nil
}

private func inferNumericProvider(_ idPrefixes: [String]) -> String? {
    var providers: [String] = []
    for prefix in idPrefixes {
        let lower = prefix.lowercased()
        let provider: String?
        if lower.hasPrefix("tmdb:") {
            provider = "tmdb"
        } else if lower.hasPrefix("trakt:") {
            provider = "trakt"
        } else if lower.hasPrefix("tvdb:") {
            provider = "tvdb"
        } else if lower.hasPrefix("simkl:") {
            provider = "simkl"
        } else {
            provider = nil
        }

        if let provider, !providers.contains(provider) {
            providers.append(provider)
        }
    }

    return providers.count == 1 ? providers[0] : nil
}

private func appendUnique(_ value: String, to values: inout [String], seen: inout Set<String>) {
    if seen.insert(value).inserted {
        values.append(value)
    }
}

private func isImdbId(_ value: String) -> Bool {
    value.range(of: "^tt\\d+$", options: [.regularExpression, .caseInsensitive]) != nil
}

private func isNumeric(_ value: String) -> Bool {
    !value.isEmpty && value.range(of: "^\\d+$", options: .regularExpression) != nil
}

private func containsSchemePrefix(_ value: String) -> Bool {
    value.range(of: "^[a-zA-Z][a-zA-Z0-9+.-]*:", options: .regularExpression) != nil
}

File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/metadata/MediaIds.kt
package com.crispy.tv.domain.metadata

data class NuvioMediaId(
    val contentId: String,
    val videoId: String?,
    val isEpisode: Boolean,
    val season: Int?,
    val episode: Int?,
    val kind: String,
    val addonLookupId: String
)

private val IMDB_ID_REGEX = Regex("^tt\\d+$", RegexOption.IGNORE_CASE)
private val NUMERIC_ID_REGEX = Regex("^\\d+$")
private val URI_SCHEME_REGEX = Regex("^[a-zA-Z][a-zA-Z0-9+.-]*:")

private data class EpisodeSuffix(
    val baseId: String,
    val season: Int?,
    val episode: Int?
)

fun normalizeNuvioMediaId(raw: String): NuvioMediaId {
    val trimmed = raw.trim()
    if (trimmed.isEmpty()) {
        return NuvioMediaId(
            contentId = "",
            videoId = null,
            isEpisode = false,
            season = null,
            episode = null,
            kind = "content",
            addonLookupId = ""
        )
    }

    val suffix = parseEpisodeSuffix(trimmed)
    if (suffix.season != null && suffix.episode != null) {
        val baseId = canonicalizeBaseId(suffix.baseId)
        if (baseId.isNotEmpty()) {
            val videoId = "$baseId:${suffix.season}:${suffix.episode}"
            return NuvioMediaId(
                contentId = baseId,
                videoId = videoId,
                isEpisode = true,
                season = suffix.season,
                episode = suffix.episode,
                kind = "episode",
                addonLookupId = videoId
            )
        }
    }

    val normalized = canonicalizeBaseId(stripSeriesPrefix(trimmed))

    return NuvioMediaId(
        contentId = normalized,
        videoId = null,
        isEpisode = false,
        season = null,
        episode = null,
        kind = "content",
        addonLookupId = normalized
    )
}

fun formatIdForIdPrefixes(
    input: String,
    mediaType: String,
    idPrefixes: List<String> = emptyList()
): String? {
    val raw = input.trim()
    if (raw.isEmpty()) {
        return null
    }

    val suffix = parseEpisodeSuffix(raw)
    val baseRaw = suffix.baseId.trim()
    if (baseRaw.isEmpty()) {
        return null
    }

    val normalizedBase = canonicalizeBaseId(baseRaw)
    val episodeSuffix = if (suffix.season != null && suffix.episode != null) {
        ":${suffix.season}:${suffix.episode}"
    } else {
        ""
    }

    val providerKind = if (mediaType.equals("movie", ignoreCase = true)) "movie" else "show"
    val candidates = linkedSetOf<String>()

    val imdbId = normalizedBase.takeIf { IMDB_ID_REGEX.matches(it) }
    val tmdbId = normalizedBase.tmdbNumericIdOrNull()

    if (imdbId != null) {
        candidates += "$imdbId$episodeSuffix"
        candidates += "imdb:$imdbId$episodeSuffix"
        candidates += "imdb:$providerKind:$imdbId$episodeSuffix"
    }

    if (tmdbId != null) {
        candidates += "tmdb:$tmdbId$episodeSuffix"
        candidates += "tmdb:$providerKind:$tmdbId$episodeSuffix"
    }

    val needsNumericInference =
        NUMERIC_ID_REGEX.matches(normalizedBase) &&
            imdbId == null &&
            tmdbId == null &&
            !URI_SCHEME_REGEX.containsMatchIn(normalizedBase)
    if (needsNumericInference) {
        val inferredProvider = inferNumericProvider(idPrefixes)
        if (inferredProvider == null && idPrefixes.isNotEmpty()) {
            return null
        }
        if (inferredProvider != null) {
            candidates += "$inferredProvider:$normalizedBase$episodeSuffix"
            candidates += "$inferredProvider:$providerKind:$normalizedBase$episodeSuffix"
        }
    }

    candidates += "$normalizedBase$episodeSuffix"

    if (idPrefixes.isNotEmpty()) {
        return candidates.firstOrNull { candidate ->
            idPrefixes.any { prefix -> candidate.startsWith(prefix) }
        }
    }

    return candidates.firstOrNull()
}

private fun stripSeriesPrefix(value: String): String {
    if (!value.startsWith("series:", ignoreCase = true)) {
        return value
    }

    return value.substring(7)
}

private fun parseEpisodeSuffix(value: String): EpisodeSuffix {
    val normalized = stripSeriesPrefix(value.trim())
    val parts = normalized.split(":")
    if (parts.size < 3) {
        return EpisodeSuffix(baseId = normalized, season = null, episode = null)
    }

    val season = parts[parts.lastIndex - 1].toIntOrNull()
    val episode = parts.last().toIntOrNull()
    if (season == null || season <= 0 || episode == null || episode <= 0) {
        return EpisodeSuffix(baseId = normalized, season = null, episode = null)
    }

    val baseId = stripSeriesPrefix(parts.dropLast(2).joinToString(":").trim())
    return EpisodeSuffix(
        baseId = baseId,
        season = season,
        episode = episode
    )
}

private fun canonicalizeBaseId(value: String): String {
    val trimmed = stripSeriesPrefix(value.trim())
    if (trimmed.isEmpty()) {
        return ""
    }

    val imdb = extractImdbId(trimmed)
    if (imdb != null) {
        return imdb
    }

    val tmdb = extractTmdbId(trimmed)
    if (tmdb != null) {
        return "tmdb:$tmdb"
    }

    return trimmed
}

private fun extractImdbId(value: String): String? {
    val normalized = stripSeriesPrefix(value.trim())
    if (IMDB_ID_REGEX.matches(normalized)) {
        return normalized.lowercase()
    }

    val parts = normalized.split(':')
    if (parts.isEmpty() || !parts.first().equals("imdb", ignoreCase = true)) {
        return null
    }

    val token = parts.asReversed().firstOrNull { part -> IMDB_ID_REGEX.matches(part) } ?: return null
    return token.lowercase()
}

private fun extractTmdbId(value: String): String? {
    val normalized = stripSeriesPrefix(value.trim())
    if (!normalized.startsWith("tmdb:", ignoreCase = true)) {
        return null
    }

    val parts = normalized.split(':')
    if (parts.size < 2) {
        return null
    }

    val direct = parts.getOrNull(1)
    if (direct != null && NUMERIC_ID_REGEX.matches(direct)) {
        return direct
    }

    val typed = parts.getOrNull(2)
    return if (typed != null && NUMERIC_ID_REGEX.matches(typed)) typed else null
}

private fun String.tmdbNumericIdOrNull(): String? {
    if (!startsWith("tmdb:", ignoreCase = true)) {
        return null
    }
    val value = substringAfter(':').substringBefore(':')
    return value.takeIf { NUMERIC_ID_REGEX.matches(it) }
}

private fun inferNumericProvider(idPrefixes: List<String>): String? {
    if (idPrefixes.isEmpty()) {
        return null
    }

    val providers =
        idPrefixes
            .mapNotNull { prefix ->
                when {
                    prefix.startsWith("tmdb:", ignoreCase = true) -> "tmdb"
                    prefix.startsWith("trakt:", ignoreCase = true) -> "trakt"
                    prefix.startsWith("tvdb:", ignoreCase = true) -> "tvdb"
                    prefix.startsWith("simkl:", ignoreCase = true) -> "simkl"
                    else -> null
                }
            }
            .distinct()

    return providers.singleOrNull()
}
